### DISCUSSION TITLE
It's difficult to compile OpenModelica by OMDev.

### DISCUSSION THREAD

**Discussion by zapplelove** [OpenModelica/OpenModelica#8482] [General]
I have followed the instruction to compile the `OMCompiler` package using the `OMDev` in windows. I have compiled `OMSimulator.exe` successfully, but when I run the command `make -f Makefile.omdev.mingw -j4 omedit`, there is no `omedit.exe` generated. And When I run the `OMSimulator.exe`, there is a message that 'can't find libstdc++-6.dll'.
Can anyone rewrite the instruction in more detail?

**Comment by adrpo** [comment]
Yes, is not easy, but it should work fine, we do it every day, several times. What readme do you use? This one?
https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/README-OMDev-MINGW.md

in the terminal, please do:
`make -f Makefile.omdev.mingw -j4 omedit > trace.txt 2>&1`
then paste the trace.txt here.

Let's see if we can find out what the issue is.

**Comment by zapplelove** [comment]
Yes, I followed the [README-OMDev-MINGW](https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/README-OMDev-MINGW.md).
Here is my operation:
In the first terminal:
```bash
cd C:\OMDev\
git checkout maintenance/v1.18
git clean -fdx
git reset --hard # In order to reset the status

.\SETUP_OMDEV.bat
.\SETUP_OMDEV_Qt5.bat
```
In the second terminal:
```bash
git checkout v1.18.0
git clean -fdx
git reset --hard # In order to reset the status

git submodule foreach --recursive git clean -fdx
git submodule foreach --recursive git reset --hard
git submodule update --init --recursive # In order to reset the status of submodules
```
In first teminal:
``` bash
.\tools\msys\mingw64.exe
```

and In the New msys teminal:
```
cd /f/Documents/Projects/OpenModelica/

export PATH=$PATH:/d/Programs/Scoop/apps/git/current/bin:/d/Programs/Scoop/apps/sliksvn/current/bin:/d/Programs/Installed/x64/Java/jdk1.8.0_311/bin
export OPENMODELICAHOME="F:\\Documents\\Projects\\OpenModelica\\build"
export OPENMODELICALIBRARY="F:\\Documents\\Projects\\OpenModelica\\build\\lib\\omlibrary"
make -f Makefile.omdev.mingw -j4 omedit > trace.txt 2>&1
```
[trace.txt](https://github.com/OpenModelica/OpenModelica/files/7535727/trace.txt)
and I have a compile error and there is no `OMedit.exe` in `F:\Documents\Projects\OpenModelica\build\bin`.

**Comment by AnHeuermann** [comment]
Works for me and that you are able to build `omc` is a good sign.

Things you could try:
* Change your shell language to English `export LANG=en_US.UTF-8` (makes it easier for us to read the log).
* Use TortoiseSVN. Not sure if that makes a different, but at least that's what I'm using and it works.
* Build OMSimulator first `make -f Makefile.omdev.mingw omsimulator -j4`.
* Your JDK seems to be from 2014(?), maybe you need to update it.

```
export LANG=en_US.UTF-8
make -f Makefile.omdev.mingw --output-sync -j4 omc &> trace_omc.txt
make -f Makefile.omdev.mingw --output-sync -j4 omsimulator &> trace_oms.txt
make -f Makefile.omdev.mingw --output-sync -j4 omedit &> trace_omedit.txt
```

By the way, I have all these exports added to C:\OMDEv\tools\msys\home\UserName\\.bashrc.
Here is it how it looks like for me:
```bash
# Use English
export LANG=en_US.UTF-8

# Use Windows path
export MSYS2_PATH_TYPE=inherit

# Constants
export OPENMODELICAHOME="d:\\workspace\\OpenModelica\\build"
export OPENMODELICALIBRARY="d:\\workspace\\OpenModelica\\build\\lib\\omlibrary"
export OMDEV="c:\\OMDev"

# Export git, svn and jdk
export PATH=$PATH:/c/Program\ Files/Git/bin:/c/Program\ Files/TortoiseSVN/bin:/c/Program\ Files/Java/jdk-15.0.1/bin

# Export OpenModelica
export PATH=$PATH:/d/workspace/OpenModelica/build/bin
```

**Comment by mahge** [comment]
I think this might be a GNU bison issue even though I am not sure why it is manifesting for @zapplelove but not for @AnHeuermann.

Can you both check your bison versions to see if there are differences.

For me I have
```sh
bison --version

bison (GNU Bison) 3.7.6
Written by Robert Corbett and Richard Stallman.

Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

```

----------

If you want a quick fix you can  try disabling building `lp_solve`. You can remove `lpsolve` from here
https://github.com/OpenModelica/OpenModelica/blob/99a01c43633c3b3d250110049468c4b548310bf3/OMCompiler/Makefile.omdev.mingw#L79

I am not sure if this will work perfectly but you give it a try until we figure out the issue.

**Comment by AnHeuermann** [comment]
>I think this might be a GNU bison issue even though I am not sure why it is manifesting for @zapplelove but not for @AnHeuermann.

In that case maybe the OMDev branch `maintenance/v1.18` isn't working with OpenModelica/OpenModelica @v1.18. Switching to the OMDev master branch could be something to try.

For me it is:
```
# bison --version
bison (GNU Bison) 3.7.2
Written by Robert Corbett and Richard Stallman.

Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by zapplelove** [comment]
@AnHeuermann 
I have tried [this](https://github.com/OpenModelica/OpenModelica/issues/8127#issuecomment-968790403) and the error message is just changed from Chinese to English.

**Comment by zapplelove** [comment]
@mahge I have switched to the OMDev master branch,  the `bison` version is `3.7.2` instead of `3.7.6`, and the error is the same.
And I tried disabling building `lp_solve`, the error message changed to 
```bash 
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[6]: *** [Makefile.common:193：/f/Documents/Projects/OpenModelica/build/lib/omc/libOpenModelicaCompiler.dll] 错误 1
make[6]: 离开目录“/f/Documents/Projects/OpenModelica/OMCompiler/Compiler/boot”
make[5]: *** [Makefile.common:70：bootstrap-from-tarball2] 错误 2
make[5]: 离开目录“/f/Documents/Projects/OpenModelica/OMCompiler/Compiler/boot”
make[4]: *** [Makefile.common:58：bootstrap-from-tarball] 错误 2
make[4]: 离开目录“/f/Documents/Projects/OpenModelica/OMCompiler/Compiler/boot”
make[3]: *** [Makefile.common:3：all] 错误 2
make[3]: 离开目录“/f/Documents/Projects/OpenModelica/OMCompiler/Compiler/boot”
make[2]: *** [Makefile.omdev.mingw:582：omc-bootstrapped] 错误 2
make[2]: 离开目录“/f/Documents/Projects/OpenModelica/OMCompiler”

real	2m8.384s
user	0m5.576s
sys	0m39.383s
make[1]: *** [Makefile.omdev.mingw:321：omc] 错误 2
make[1]: 离开目录“/f/Documents/Projects/OpenModelica/OMCompiler”
make: *** [Makefile.omdev.mingw:65：omc] 错误 2
```
[trace.txt](https://github.com/OpenModelica/OpenModelica/files/7542717/trace.txt)

**Comment by adrpo** [comment]
The error is not the same. Now there is an error when linking lpsolve to libOpenModeilcaCompiler.dll as it does not exist.
Please put back the build of lpsolve:
```
cd OpenModelica/OMCompiler/
git checkout -- Makefile.omdev.mingw
```
then try to build again.

**Comment by zapplelove** [comment]
> The error is not the same. Now there is an error when linking lpsolve to libOpenModeilcaCompiler.dll as it does not exist. Please put back the build of lpsolve:
> 
> ```
> cd OpenModelica/OMCompiler/
> git checkout -- Makefile.omdev.mingw
> ```
> 
> then try to build again.

This will result in the first error, the lpsolve package compilation error.

**Comment by adrpo** [comment]
I can't see that, which trace is it?

**Comment by zapplelove** [comment]
[This error](https://github.com/OpenModelica/OpenModelica/files/7535727/trace.txt):
```bash
......
......
make[4]: 进入目录“/f/Documents/Projects/OpenModelica/OMCompiler/3rdParty/libffi/x86_64-w64-mingw32”
 /usr/bin/install -c yacc_read.h /f/Documents/Projects/OpenModelica/OMCompiler/3rdParty/lpsolve/build/include/lpsolve/
[100%] Built target minizip
make[3]: 离开目录“/f/Documents/Projects/OpenModelica/OMSimulator/3rdParty/zlib/build/mingw/minizip”
bison --no-lines -y -d lp_rlp.y
lp_rlp.y:16.1-12: 警告: POSIX Yacc does not support %pure-parser [-Wyacc]
   16 | %pure-parser
      | ^~~~~~~~~~~~
lp_rlp.y:16.1-12: 警告: 已弃用的指令：“%pure-parser”，应使用 “%define api.pure” [-Wdeprecated]
   16 | %pure-parser
      | ^~~~~~~~~~~~
      | %define api.pure
Making all in include
Install the project...
make[5]: 进入目录“/f/Documents/Projects/OpenModelica/OMCompiler/3rdParty/libffi/x86_64-w64-mingw32/include”
make[5]: 对“all”无需做任何事。
make[5]: 离开目录“/f/Documents/Projects/OpenModelica/OMCompiler/3rdParty/libffi/x86_64-w64-mingw32/include”
-- Install configuration: ""
-- Up-to-date: F:/Documents/Projects/OpenModelica/OMSimulator/3rdParty/zlib/install/mingw/lib/libminizip.a
-- Up-to-date: F:/Documents/Projects/OpenModelica/OMSimulator/3rdParty/zlib/install/mingw/include/minizip.h
-- Up-to-date: F:/Documents/Projects/OpenModelica/OMSimulator/3rdParty/zlib/install/mingw/include/miniunz.h
Making all in testsuite
lp_rlp.y: 致命错误: invalid @ in skeleton: @
make[2]: 离开目录“/f/Documents/Projects/OpenModelica/OMSimulator/3rdParty/zlib/build/mingw/minizip”
make[1]: 离开目录“/f/Documents/Projects/OpenModelica/OMSimulator”
make -C OMSimulator config-OMSimulator OMBUILDDIR=/f/Documents/Projects/OpenModelica/build CC="clang" CXX="clang++"
make[2]: *** [Makefile:91：lp_rlp.c] 错误 1
make[2]: 离开目录“/f/Documents/Projects/OpenModelica/OMCompiler/3rdParty/lpsolve”
make[1]: *** [Makefile.common:581：/f/Documents/Projects/OpenModelica/build/lib/omc/liblpsolve55.a] 错误 2
make[1]: *** 正在等待未完成的任务....
......
......
```
@adrpo

**Comment by adrpo** [comment]
can you do:
`which bison`
in the terminal after you set the paths?

**Comment by zapplelove** [comment]
> can you do: `which bison` in the terminal after you set the paths?

![image](https://user-images.githubusercontent.com/18183148/141953113-01260e91-d796-4645-a91a-0a34c34a5460.png)
![image](https://user-images.githubusercontent.com/18183148/141953255-544ceca6-c12b-4384-9226-f3948fe61e39.png)

**Comment by adrpo** [comment]
Try this:
```
cd OpenModelica/OMCompiler/3rdParty/lpsolve
# clean the thing so we make sure new files are generated
git clean -fdx 
bison --no-lines -y -d lp_rlp.y
```

**Comment by zapplelove** [comment]
![image](https://user-images.githubusercontent.com/18183148/141956228-edc1eb84-d816-4a60-9f35-b67988e67e3d.png)

**Comment by adrpo** [comment]
The only thing I can think of is you did not force Unix line endings before cloning OMDev:
```
  git config --global core.eol lf
  git config --global core.autocrlf input
```
And because of that the line endings in some of the text files are CRLF instead of LF as expected and some tools really don't like that.

**Comment by AnHeuermann** [comment]
I could try to set up OMDev on an new Windows machine sometime this week and check if I do something different. I'll need to do it at some point anyway.

An alternative could be to use Windows Subsystem for Linux, if that is an option for @zapplelove. It is more easy and faster, but of course "only" Linux.

**Comment by AnHeuermann** [comment]
Here is what I did on Windows to build OpenModelica with OMDev. You'll need to update the paths to match your system.
Also I tried to build the master first (I know that that one is working).

# Building OpenModelica on Windows 10

## Install git

Use the recommendations when installing git.
(Some git tools could clash with Windows or OMDev tools.)

## Checkout OMDev with git

Install OMDev in `C:\OMDev`. Other locations could cause trouble.

Run in git-bash:
```bash
$ git config --global core.eol lf
$ git config --global core.autocrlf input
$ cd /c/
$ git clone https://openmodelica.org/git/OMDev.git
```

## Set up OMDev


- Define OMDEV environment variable pointing to `C:\OMDev`.
  1. For Windows 10, right click on the Start Menu and select System.
     Click System Info.
     click advanced system settings.
  2. Select Advanced system settings.
  3. In the Advanced tab, select Environment Variables.
  4. Select New and add `OMDEV` pointing to `C:\OMDev`.

- Setup OMDEV

  Run in Windows CMD:
  ```cmd
  C:\>cd OMDev
  C:\OMDev>SETUP_OMDEV.bat
  C:\OMDev>SETUP_OMDEV.bat
  ```

- Start `C:\OMDev\tools\msysmingw64` for the first time to create user files.

  ```bash
  mkdir: cannot change permissions of '/dev/shm': Permission denied
  mkdir: cannot change permissions of '/dev/mqueue': Permission denied
  Copying skeleton files.
  These files are for the users to personalise their msys2 experience.

  They will never be overwritten nor automatically updated.

  './.bashrc' -> '/home/username/.bashrc'
  './.bash_logout' -> '/home/username/.bash_logout'
  './.bash_profile' -> '/home/username/.bash_profile'
  './.inputrc' -> '/home/username/.inputrc'
  './.profile' -> '/home/username/.profile'
  './.vimrc' -> '/home/username/.vimrc'

  username@PC-name MINGW64 ~
  ```

## Installing JDK and SVN

- Install latest Java SE Development Kit
- Install latest Tortoise SVN

### Set up .bashrc

- Add the following to the end of `C:\OMDev\tools\msys\home\username\.bashrc`.
  Make sure to use the correct paths for your system.

  ```bash
  # Use Windows path
  export MSYS2_PATH_TYPE=inherit

  # Constants
  export OPENMODELICAHOME="d:\\workspace\\OpenModelica\\build"
  export OPENMODELICALIBRARY="d:\\workspace\\OpenModelica\\build\\lib\\omlibrary"
  export OMDEV="c:\\OMDev"

  # Export git, svn and jdk
  export PATH=$PATH:/c/Program\ Files/Git/bin:/c/Program\ Files/TortoiseSVN/bin:/c/Program\ Files/Java/jdk-17.0.1/bin

  # Export OpenModelica
  export PATH=$PATH:/d/workspace/OpenModelica/build/bin
  ```
- Restart mingw64-bash and check if svn, git and javac can be found:
  ```bash
  # which svn
  # which git
  # which javac
  ```

## Download OpenModelica

Make sure you don't try to build OpenModelica on a network drive.

Run in git-bash:
```bash
$ cd /d/workspace/
$ git clone https://github.com/OpenModelica/OpenModelica.git --recursive
```

## Build OpenModelica

Run in mingw64-bash:
```bash
# cd /d/workspace/OpenModelica/
# make -f Makefile.omdev.mingw -j6
```

( Don't use all threads if you don't have a large amount of memory. On my laptop I usually need to use a thread less.)

**Comment by zapplelove** [comment]
I think the error caused by this.
> git config --global core.eol lf
> git config --global core.autocrlf input

thanks!

**Comment by AnHeuermann** [comment]
Anything we could improve in our README? What I summarized in https://github.com/OpenModelica/OpenModelica/issues/8127#issuecomment-975437743 is basically our README.

I'm afraid we just have to accept, that OpenModelica is complicated to build on Windows.
So should we close this issue?

**Comment by casella** [comment]
@AnHeuermann what you wrote in your comment seems pretty informative. If it stays where it is, it is pretty useless, as I'm afraid robot and crawlers scan forums but not our tickets. Is there a README for building omc on Windows? Could you put it there?

**Comment by AnHeuermann** [comment]
> Is there a README for building omc on Windows? Could you put it there?

Well, it is simply the content of https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/README-OMDev-MINGW.md and what's in OMDev\INSTALL.txt, just for one specific case. The original README is a bit more flexible then what I did.

The stuff in OMDev\INSTALL.txt should stay in there, but maybe we should make it a markdown document and restructure it a bit. But that should happen in the OMDev repo. Where is that any way? Can't find it on GitHub.

**Comment by casella** [comment]
@adrpo can you shed some light on the whereabouts of the OMDev repo?

**Comment by sjoelund** [comment]
https://openmodelica.org/git/OMDev.git

**Comment by adrpo** [comment]
It says in the https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/README-OMDev-MINGW.md 
`git clone https://openmodelica.org/git/OMDev.git` but maybe that's just too hidden :)

**Comment by AnHeuermann** [comment]
Would making OMDev a repo on GitHub, with README and all, be a good idea?
But I guess there is not much value in it. There aren't that many people building on Windows and even less would want to change anything in OMDev.

**Comment by adrpo** [comment]
And its size is huge that's why we put it in opemodelica.org. 
We had OMDev in github before OpenModelica was accepted as a non-profit organization and the max size of a repo was less than 2GB.

**Comment by zapplelove** [comment]
@adrpo Any suggestions about compiling OMEdit using `cmake` method?

I find that there is a `CMakeLists.txt` in the root folder, so I use following command to compile the project in Ubuntu:
```bash
mkdir build & cd build
cmake .. -DCMAKE_INSTALL_PREFIX=~/local/Openmodelica
make -j16 
```

and  I don't find the command `make omedit`

**Comment by mahge** [comment]
> and I don't find the command make omedit

@zapplelove. Unfortunately the CMake configuration does not include OMEdit yet. It will come soon. 

For now you would have to use the `Makefile.omdev.mingw` to build OMEdit.

**Comment by zapplelove** [comment]
@mahge Ok, I think the `CMake` compile method is more friendly for me to understand the whole project and relationships between each submodules. I want to understand how a `metamodelica file` being translated to a `c file` and compiled to a `bin file`. And the `GNU autoconf` and `Makefile` is diffucult for me. I want to recompile the project to work with other simulation programs like drake, gazebo, and so on, instead of simulink.

**Comment by mahge** [comment]
I understand. Some context about the different parts of OpenModelica first. `OMEdit` is the graphical user interface and editor for the OpenModelica ecosystem. It is written completely in C++ with Qt. It does not handle any Modelica "compilation" itself.

The actual Modelica code compilation is handled by `omc`, the OpenModelica Modelica Compiler. This is where `MetaModelica` and a bunch of other languages are used. 

So if your interest is mainly in the "compiler" part of the ecosystem, then what you want to focus on is `omc` itself and not `OMEdit`.

Right now the `CMake` support is in development and does not cover the whole of OpenModelica. At the moment it covers everything in the `OMCompiler` folder so you can compile `omc` with it (`make omc`). However, I would not recommend it for general usage for someone new to OpenModelica yet. 

You can find the bulk of MetaModelica compilation logic in `OMCompiler/Compiler/CMakeLists.txt`. A word of warning though, it is not the most straight forward CMake operation, especially if you are not vary familiar with OpenModelica yet.
